src/vsr/journal.zig

```diff
@@ -221,73 +248,126 @@ pub fn Journal(comptime Replica: type, comptime Storage: type) type {
         }
 
         /// Asserts that headers are .reserved (zeroed) from `op_min` (inclusive).
-        pub fn assert_headers_reserved_from(self: *Self, op_min: u64) void {
+        pub fn assert_headers_reserved_from(self: *const Self, op_min: u64) void {
             // TODO Snapshots
-            for (self.headers[op_min..]) |header| assert(header.command == .reserved);
+            for (self.headers) |*header| {
+                assert(header.command == .reserved or header.op < op_min);
+            }
+        }
+
+        /// Returns whether this is a fresh database WAL; no prepares (except the init) have ever
+        /// been written.
+        pub fn is_empty(self: *const Self) bool {
+            if (self.headers[0].operation != .init) return false;
+
+            // If the message immediately following the initial prepare is damaged, we must fall
+            // back to VSR recovery protocol (i.e. treat this as a non-empty WAL) since that
+            // message may have been a prepare.
+            if (self.faulty.bit(Slot{ .index = 1 })) return false;
+
+            // Other headers are allowed to be damaged (they will be repaired from status=normal)
+            // but there shouldn't be any prepares.
+            for (self.headers[1..]) |*header| {
+                if (header.command == .prepare) return false;
+            }
+
+            return true;
```

Comment:
Can we make this stronger still, before returning true? i.e. Do we expect any dirty/faulty bits to be set at all?

Is there anything else we could possibly assert to be absolutely sure that the journal is empty?

Reply:
> Do we expect any dirty/faulty bits to be set at all?

Right now dirty/faulty bits _can_ be set, though it is a little tricky. The criteria I use for an "empty" journal:

- slot 0 is the initial prepare, and
- slot 1 is valid and reserved, and
- all other slots are either reserved or faulty (the faulty slots are automatically cleared during recovery, since `replica.op < slot_count` and `op_known` and `faulty_op > replica.op`).

It is possible that one of those faulty ops might have been an actual message, in which case we would do the wrong thing. The alternative is that any corruption in a WAL automatically makes it "non-empty" and thus require recovery protocol. What do you think?

---

> Is there anything else we could possibly assert to be absolutely sure that the journal is empty?

- Assert that the journal has recovered, since `is_empty` cannot be safely called prior to that.
- Assert the `init` prepare's checksum is as expected.
- Verify that `prepare_checksums[1..]` are each `0`.

Reply:
Thanks for providing the criteria, that's really clear.

> It is possible that one of those faulty ops might have been an actual message, in which case we would do the wrong thing.

Yes, that's a real possibility. That kind of scenario also fits better how we really think of prepares, which is that they can be prepared out of order, e.g. slot 2 before slot 1.

> The alternative is that any corruption in a WAL automatically makes it "non-empty" and thus require recovery protocol. What do you think?

Let's do somewhere in between these two strategies, where we modify the 2nd criteria to require all the slots for the pipeline to all be clean and reserved, but any slots beyond can be reserved or faulty. That should do the trick right?

Reply:
> Let's do somewhere in between these two strategies, where we modify the 2nd criteria to require all the slots for the pipeline to all be clean and reserved, but any slots beyond can be reserved or faulty. That should do the trick right?

What about:
- `replica_count=3` and `journal_slot_count=128`.
- Replicas `0` and `1` are healthy & processing messages.
- A network partition prevents any prepares from reaching replica `2` until e.g. op 50.
- Then replica `2` crashes and restarts.

If replica `2`'s op 50 is corrupt, it would start up as an empty replica.

Reply:
Implemented in https://github.com/coilhq/tigerbeetle/pull/91/commits/0d757910771f88d72d754d9c0596af561e701320.